File: adjplaces.pl
#!/usr/bin/env perl
use strict;

use DBI;
use Getopt::Long;


# Array of table/field hashes.
my @dbpieces = (
   { table => "tng_families", field => "marrplace" },
   { table => "tng_families", field => "divplace" },
   { table => "tng_people", field => "birthplace" },
   { table => "tng_people", field => "altbirthplace" },
   { table => "tng_people", field => "deathplace" },
   { table => "tng_people", field => "burialplace" },
   { table => "tng_events", field => "eventplace" },
   { table => "tng_places", field => "place" },
);

my @states = split( /\s\s+/,
   q(Alabama
   Alaska
   Arizona
   Arkansas
   California
   Colorado
   Connecticut
   Delaware
   District of Columbia
   Florida
   Georgia
   Hawaii
   Idaho
   Illinois
   Indiana
   Iowa
   Kansas
   Kentucky
   Louisiana
   Maine
   Maryland
   Massachusetts
   Michigan
   Minnesota
   Mississippi
   Missouri
   Montana
   Nebraska
   Nevada
   New Hampshire
   New Jersey
   New Mexico
   New York
   North Carolina
   North Dakota
   Ohio
   Oklahoma
   Oregon
   Pennsylvania
   Rhode Island
   South Carolina
   South Dakota
   Tennessee
   Texas
   Utah
   Vermont
   Virginia
   Washington
   West Virginia
   Wisconsin
   Wyoming)
);

# Getopt variables
my ($db_host, $db_name, $db_user, $db_pw, $doupdate);

GetOptions(
   'h|host=s'     => \$db_host,
   'd|database=s' => \$db_name,
   'u|user=s'     => \$db_user,
   'update'       => \$doupdate,
   'p|password=s' => \$db_pw,
);

# Connect to the database using the supplied information.
my $dbh = DBI->connect("dbi:mysql:$db_name:$db_host", "$db_user", "$db_pw", {RaiseError => 1, PrintError => 1}) or die("Cannot connect to the database: " . DBI->errstr);

# Loop through each state
foreach my $state (@states) {
   print "Searching for $state..\n";
   # Then loop through each table/field pair
   foreach my $dbpiece (@dbpieces) {
      my $table = $dbpiece->{'table'};
      my $field = $dbpiece->{'field'};
      print "  in $table > $field: ";
      if ($doupdate ) {
         my $sql = "UPDATE $table SET $field = replace($field, '$state', '$state, USA') WHERE $field REGEXP '$state\$'";
         my $sth = $dbh->prepare( $sql );
         $sth->execute();
         print "(" . $sth->rows . ") rows affected.\n";
      }
      else {
         my $sql = "SELECT * FROM $table WHERE $field REGEXP '$state\$'";
         my $sth = $dbh->prepare( $sql );
         $sth->execute();
         #my @results = $sth->fetchrow_array();
         print "(" . $sth->rows . ") rows found.\n";
      }

   }
}

$dbh->disconnect or die("Cannot disconnect from the database");

File: adjplaces.php
<html>
<head></head>
<body>
<?php
include( "tng_begin.php" );

#ini_set('display_errors',1);
#error_reporting(E_ALL);

$dbpieces = array(
   array( table => "tng_families", field => "marrplace" ),
   array( table => "tng_families", field => "divplace" ),
   array( table => "tng_people", field => "birthplace" ),
   array( table => "tng_people", field => "altbirthplace" ),
   array( table => "tng_people", field => "deathplace" ),
   array( table => "tng_people", field => "burialplace" ),
   array( table => "tng_events", field => "eventplace" ),
   array( table => "tng_places", field => "place" ),
);

$states = array(
   "Alabama",
   "Alaska",
   "Arizona",
   "Arkansas",
   "California",
   "Colorado",
   "Connecticut",
   "Delaware",
   "District of Columbia",
   "Florida",
   "Georgia",
   "Hawaii",
   "Idaho",
   "Illinois",
   "Indiana",
   "Iowa",
   "Kansas",
   "Kentucky",
   "Louisiana",
   "Maine",
   "Maryland",
   "Massachusetts",
   "Michigan",
   "Minnesota",
   "Mississippi",
   "Missouri",
   "Montana",
   "Nebraska",
   "Nevada",
   "New Hampshire",
   "New Jersey",
   "New Mexico",
   "New York",
   "North Carolina",
   "North Dakota",
   "Ohio",
   "Oklahoma",
   "Oregon",
   "Pennsylvania",
   "Rhode Island",
   "South Carolina",
   "South Dakota",
   "Tennessee",
   "Texas",
   "Utah",
   "Vermont",
   "Virginia",
   "Washington",
   "West Virginia",
   "Wisconsin",
   "Wyoming",
);

foreach ($states as $state) {
   # What we are searching for.  Note the SQL statement anchors the search to
   # the end of the line. Adjust that too if need be.
   $searchstr = "$state";
   # What we replace it with
   $replacestr = "$state, USA";
   echo "<p>Searching for '$searchstr\$'...<br />\n";
   foreach ($dbpieces as $dbpiece) {
      $table = $dbpiece['table'];
      $field = $dbpiece['field'];
      echo "  $table > $field = ";
      # If the script is passed 'update=1', i.e. adjplace.php?update=1.
      if ($_GET['update']) {
         # The regex '$' only matches at the end of the line.
         $sql = "UPDATE $table SET $field = replace($field, '$searchstr', '$replacestr') WHERE $field REGEXP '$searchstr\$'";
         $result = mysql_query( $sql ) or die ( $text['cannotexecutequery'] . ": $sql" );
         $num = mysql_affected_rows( $result );
         echo "$num rows updated.<br />\n";
      }
      # Otherwise just query the database.
      else {
         $sql = "SELECT * FROM $table WHERE $field REGEXP '$searchstr\$'";
         $result = mysql_query( $sql ) or die ( $text['cannotexecutequery'] . ": $sql" );
         $num = mysql_num_rows( $result );
         echo "$num rows found.<br />\n";
         # If the script is passed 'details=1', print the place field for each record.
         if ($_GET['details']) {
            while ($row = mysql_fetch_assoc($result)) {
               echo "    {$row[$field]}<br />";
            }
         }
      }
   }
   echo "</p>\n";
}

?>
</body>
</html>
